metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,339 @@
+package io.iohk.metronome.networking
+
+import cats.effect.concurrent.Deferred
+import cats.effect.implicits._
+import cats.effect.{Concurrent, ContextShift, Resource, Sync, Timer}
+import cats.implicits._
+import io.iohk.metronome.networking.ConnectionHandler.HandledConnection
+import io.iohk.metronome.networking.EncryptedConnectionProvider.ConnectionError
+import io.iohk.metronome.networking.RemoteConnectionManager.{
+  ConnectionAlreadyClosedException,
+  MessageReceived
+}
+import monix.catnap.ConcurrentQueue
+import monix.eval.{TaskLift, TaskLike}
+import monix.reactive.Observable
+import monix.tail.Iterant
+import scodec.Codec
+
+import java.net.InetSocketAddress
+import java.util.concurrent.{ThreadLocalRandom, TimeUnit}
+import scala.concurrent.duration.FiniteDuration
+
+class RemoteConnectionManager[F[_]: Sync, K, M: Codec](
+    connectionHandler: ConnectionHandler[F, K, M],
+    localInfo: (K, InetSocketAddress)
+) {
+
+  def getLocalInfo: (K, InetSocketAddress) = localInfo
+
+  def getAcquiredConnections: F[Set[K]] = {
+    connectionHandler.getAllActiveConnections
+  }
+
+  def incomingMessages: Iterant[F, MessageReceived[K, M]] =
+    connectionHandler.incomingMessages
+
+  def sendMessage(recipient: K, message: M): F[Unit] = {
+    connectionHandler.getConnection(recipient).flatMap {
+      case Some(connection) =>
+        //Connections could be closed by remote without us noticing, close it on our side and return error to caller
+        connection.sendMessage(message).handleErrorWith { e =>
+          //Todo logging
+          connection
+            .close()
+            .flatMap(_ =>
+              Sync[F].raiseError(ConnectionAlreadyClosedException(recipient))
+            )
+        }
```

Comment:
This looks to me like it should live in the `ConnectionHandler`, because this code calls `close` on the `HandledConnection` itself, but I think it should instead call `ConnectionHandler.closeAndDeregisterConnection`, otherwise this dead connection for example prevents incoming ones from taking its place. I'm not even sure how you learn that it should be replaced with a new one?

I reckon that the connection handler shouldn't return a `HandledConnection` that you can close without it being aware of it. Either it should return a `Resource[Option[Kleisli[F, M, Unit]]]` that you can use to send a message, but nothing else, or it should have a `sendMessage` method of its own returning an `EitherT[F, ConnectionAlreadyClosed, Unit]` and make sure that the proper callbacks are called, bookkeeping is done, etc. I thought about a `withConnection` kind of method, but because it has to receive an `Option` it's not as clean as just moving `sendMessage` in.

Also, maybe it would be worth checking that `e` here is really corresponds to a closed connection, not something unexpected, maybe related to the serialisation of `M`?

Reply:
Totally agree ! To be honest the more I think about it, the more i think that `HandledConnection` should be just internal representation of `ConnectionHandler`